src/System.Slices/System/Buffers/Memory.cs

```diff
@@ -0,0 +1,132 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Represents a Span<byte> factory</byte>
+    /// </summary>
+    /// <remarks>
+    /// This struct is not safe for multithreaded access, i.e. it's subject to struct tearing that can result in unsafe memory access. 
+    /// </remarks>
+    public struct Memory<T>
+    {
+        public static unsafe Memory<T> Empty = new Memory<T>(null, 0, 0);
+
+        private readonly T[] _array;
+        private readonly int _offset;
+        private readonly unsafe void* _memory;
+        private readonly int _memoryLength;
+
+        public unsafe Memory(void* pointer, int length)
+        {
+            unsafe
+            {
+                _memory = pointer;
+            }
+
+            _array = null;
+            _offset = 0;
+            _memoryLength = length;
+        }
+
+        public unsafe Memory(T[] array) : this(array, 0, array.Length)
+        {
+        }
+
+        public unsafe Memory(T[] array, int offset, int length, bool isPinned = false)
+        {
+            unsafe
+            {
+                if (isPinned)
+                {
+                    // The caller pinned the array so we can safely get the pointer
+                    _memory = (void*)UnsafeUtilities.ComputeAddress(array, new UIntPtr(0));
+                }
+                else
+                {
+                    _memory = null;
+                }
+            }
+
+            _array = array;
+            _offset = offset;
+            _memoryLength = length;
+        }
+
+        public Span<T> Span
```

Comment:
We need this property, casting is ugly. I started without it and added it because of the uglyness. Extension methods are one of the big reasons you want this, also it's unclear when code starts to do this:

```C#
Span<byte> span = Memory;
```

All of the variable declarations go from `var` blah to `Span<T>` to force implicit conversions to work:

```C#
var span = Memory.Span;
```